### PR TITLE
fix(tianmu): Solve the problem of query failure when a=a occurs in the predicate or. (#301, #733)

### LINKED ISSUE #301: bug: SELECT * FROM t1 WHERE (a=a AND a=a) OR b > 2;No data can be queried
### Describe the problem

```
CREATE TABLE t1 (a INT NOT NULL, b INT)engine=tianmu;
INSERT INTO t1 VALUES (1, 1);
SELECT * FROM t1 WHERE (a=a AND a=a) OR b > 2;
##The error results are as follows
Empty set (0.01 sec)
```

### Expected behavior

```
##The following is：innodb test result
SELECT * FROM t1 WHERE (a=a AND a=a) OR b > 2;
+---+------+
| a | b    |
+---+------+
| 1 |    1 |
+---+------+
1 row in set (0.00 sec)
```

### How To Reproduce

```
CREATE TABLE t1 (a INT NOT NULL, b INT)engine=stonedb;
INSERT INTO t1 VALUES (1, 1);
SELECT * FROM t1 WHERE (a=a AND a=a) OR b > 2;
```

### Environment

_No response_

### Are you interested in submitting a PR to solve the problem?

- [ ] Yes, I will!

### PR DESCRIPTION
**Pull request by adofsauron**
## Summary about this PR

    The reason for the error lies in that in Item_cond_or condition,
    sub-conditions similar to a=a will be cleared after
    elimination of equivalent conditions,
    but the tianmu engine does not do additional scene
    processing for the remaining conditions after elimination
    of equivalent conditions of Item_cond_or.
    Results in a scenario where
    only the remaining predicates are processed
    and the or predicates are not considered.

    The corresponding issuse:
    https://github.com/stoneatom/stonedb/issues/301
    https://github.com/stoneatom/stonedb/issues/733
Issue Number: close #301  #733 

## Tests Check List

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

## Changelog

- [ ] New Feature
- [x] Bug Fix
- [ ] Improvement
- [ ] Performance Improvement
- [ ] Build/Testing/CI/CD
- [ ] Documentation
- [ ] Not for changelog (changelog entry is not required)

## Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features

### CODE REVIEW

File: sql/sql_optimizer.cc [hunk 1]
```diff
@@ -10157,8 +10163,18 @@ static bool internal_remove_eq_conds(THD *thd, Item *cond,
       if (internal_remove_eq_conds(thd, item, &new_item, &tmp_cond_value, part))
         return true;
 
-      if (new_item == NULL)
-        li.remove();
+			if (new_item == NULL)
```
Comment (RingsC) [inline sql/sql_optimizer.cc]: using nullptr

**Review by RingsC — CHANGES_REQUESTED** [review]

File: sql/sql_optimizer.cc [hunk 2]
```diff
@@ -10157,8 +10163,18 @@ static bool internal_remove_eq_conds(THD *thd, Item *cond,
...
+			{
+				if (is_cond_or)
+				{
+					Item* cond_true = new Item_int((longlong)1, 1);  // Always true
```
Comment (RingsC) [inline sql/sql_optimizer.cc]: using new placement, mem_root

**Review by RingsC — CHANGES_REQUESTED** [review]

File: sql/sql_optimizer.cc [hunk 3]
```diff
@@ -10141,6 +10141,12 @@ static bool internal_remove_eq_conds(THD *thd, Item *cond,
                                      Item::cond_result *cond_value,
                                      unsigned char part)
 {
+	bool is_cond_or = false;
+	if (dynamic_cast<Item_cond_or*>(cond))
```
Comment (RingsC) [inline sql/sql_optimizer.cc]: using `if (cond && dynamic_cast<Item_cond_or*>(cond))`

**Review by RingsC — CHANGES_REQUESTED** [review]

File: sql/sql_optimizer.cc (hunk 3 above)
Comment (RingsC) [reply sql/sql_optimizer.cc]: bool is_cond_or = (cond && dynamic_cast<Item_cond_or*>(cond)) ? ture : false

**Review by RingsC — CHANGES_REQUESTED** [review]

File: sql/sql_optimizer.cc (hunk 2 above)
Comment (adofsauron) [reply sql/sql_optimizer.cc]: yes, wait a memont for fix this

File: sql/sql_optimizer.cc (hunk 3 above)
Comment (adofsauron) [reply sql/sql_optimizer.cc]: sure,that's gook

File: sql/sql_optimizer.cc (hunk 1 above)
Comment (adofsauron) [reply sql/sql_optimizer.cc]: If the original code is modified, it will break the whole verification rule, right?

**Review by RingsC — APPROVED** [review]
LGTM

**Review by lujiashun — APPROVED** [review]
LGTM

**Pull request merged**